app/enterprise.go

```diff
@@ -136,12 +142,20 @@ func (s *Server) initEnterprise() {
 		s.Notification = notificationInterface(s.FakeApp())
 	}
 	if samlInterface != nil {
-		s.Saml = samlInterface(s.FakeApp())
-		s.AddConfigListener(func(_, cfg *model.Config) {
+		setSAMLInterface := func(_, cfg *model.Config) {
+			if *cfg.ExperimentalSettings.UseNewSAMLLibrary && samlInterfaceNew != nil {
+				mlog.Debug("Loading new SAML2 library")
+				s.Saml = samlInterfaceNew(s.FakeApp())
+			} else {
+				mlog.Debug("Loading original SAML library")
+				s.Saml = samlInterface(s.FakeApp())
```

Comment:
In the default case, I believe every configuration change will overwrite `s.Saml` with a new pointer. Go doesn't guarantee atomicity here, so it seems like a concurrent reader might panic on a corrupted read. Also, creating new instances all the time seems wasteful -- though unlikely to have any meaningful performance overhead.

One option would be to guard access to `s.Saml` with a `RWMutex` throughout the application, but that itself seems painful. Is there any chance of delegating the choice of implementation into the existing`Saml` instance, abstracting the necessary locking there instead?